components/google-cloud/storage/upload_dir_contents/component.yaml

```diff
@@ -0,0 +1,32 @@
+name: Upload to GCS
```

Comment:
What is the difference between this component and upload_to_explicit_uri?

Reply:
Upload to explicit URI can be used to control the directory name, but it still uploads the directory. Suppose we have a folder `tmp` containing files `1`, `2`, `3` and we upload it to `gs://my-bucket/some-folder`. `upload_to_explicit_uri` creates:

```
my-bucket
  some-folder
    tmp
      1
      2
      3
```

while `upload_dir_contents` would give:

```
my-bucket
  some-folder
    1
    2
    3
```

Reply:
I think this might be in scope of upload_to_explicit_uri

>Suppose we have a folder tmp containing files 1, 2, 3
What is supposed to happen if there are two directories - tmp1 and tmp2? Where does tmp2 go?

The expected behavior of upload_to_explicit_uri

Given a file artifact:
/inputs/some_input/data
should be uploaded to
gcs_path

Given a directory artifact:
/inputs/some_input/data/*
should be uploaded to
gcs_path/*

If the behavior is different, it's a bug.

Which can be summarized as:

/inputs/some_input/data{any_suffix} --> gcs_path{any_suffix}

Reply:
Sorry for being late to this, I've been busy with other work.

Ah, so the behavior should be different based on the presence or absence of `/*`? Without that suffix, I think we see the duplicate directories owing to the use of `gsutil cp` rather than `gsutil rsync`. There's an asymmetry here that is confusing: the download components use `rsync` rather than `cp` to avoid this behavior (that's explicitly documented in the component source). As a user, I would expect upload and download components to behave the same way.

If appending `/*` is the expected way to address this, and I'm feeding paths from upstream components, is it expected that those components should change? Is there a way to append this if I don't have control over those components? I think using something like `f"{task.output}/*.*"` discards anything following the variable name. If I was writing my own component I could use a `concat` placeholder...